Battle of the Neighbourhoods - Toronto

In [6]:
import pandas as pd

Question 1 - Importing Data


In [7]:
import requests

In [8]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [9]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [10]:
len(wiki_data), type(wiki_data)

(3, list)

In [11]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [12]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [13]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [14]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [15]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [16]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [17]:
df.shape

(103, 3)

Answer to Question 1 is 103 rows and 3 columns

Question 2 - Installing Geocoder

In [18]:
pip install geocoder

     |████████████████████████████████| 98 kB 10.1 MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [19]:
import geocoder

In [20]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [21]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", data.shape)


The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


In [22]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [23]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [24]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [25]:
combined_data.shape

(103, 5)

Answer to Question 2 is 103 rows and 5 columns

Question 3 - Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [1]:
from geopy.geocoders import Nominatim

In [2]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [4]:
!conda install -c conda-forge folium=0.5.0 --yes
import folium

Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python-3.7-main

  added / updated specs:
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    _libgcc_mutex-0.1          |      conda_forge           3 KB  conda-forge
    _openmp_mutex-4.5          |           1_llvm           5 KB  conda-forge
    _py-xgboost-mutex-2.0      |            cpu_0           8 KB  conda-forge
    _pytorch_select-0.2        |            gpu_0           2 KB
    absl-py-0.11.0             |   py37h89c1867_0         168 KB  conda-forge
    aiohttp-3.7.4              |   py37h5e8e339_0         632 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
   

  future                    pkgs/main::future-0.18.2-py37_1 --> conda-forge::future-0.18.2-py37h89c1867_3
  gast                pkgs/main/linux-64::gast-0.2.2-py37_0 --> conda-forge/noarch::gast-0.4.0-pyh9f0ad1d_0
  glib                    pkgs/main::glib-2.65.0-h3eb4bd4_0 --> conda-forge::glib-2.66.7-h9c3ff4c_1
  gmp                       pkgs/main::gmp-6.1.2-h6c8ec71_1 --> conda-forge::gmp-6.2.1-h58526e2_0
  google-auth            pkgs/main::google-auth-1.22.0-py_0 --> conda-forge::google-auth-1.24.0-pyhd3deb0d_0
  google-auth-oauth~ pkgs/main::google-auth-oauthlib-0.4.1~ --> conda-forge::google-auth-oauthlib-0.4.2-pyhd8ed1ab_0
  grpcio             pkgs/main::grpcio-1.27.2-py37hf8bcb03~ --> conda-forge::grpcio-1.36.1-py37hb27c1af_0
  gst-plugins-base   pkgs/main::gst-plugins-base-1.14.0-hb~ --> conda-forge::gst-plugins-base-1.18.3-h04508c2_0
  gstreamer          pkgs/main::gstreamer-1.14.0-hb31296c_0 --> conda-forge::gstreamer-1.18.3-h3560a44_0
  h5py                pkgs/main::h5py-2

scikit-image-0.18.1  | 11.5 MB   | ##################################### | 100% 
sleef-3.5.1          | 1.5 MB    | ##################################### | 100% 
sqlite-3.34.0        | 1.4 MB    | ##################################### | 100% 
dbus-1.13.6          | 585 KB    | ##################################### | 100% 
mock-4.0.3           | 51 KB     | ##################################### | 100% 
nss-3.62             | 2.1 MB    | ##################################### | 100% 
mkl-2020.4           | 215.6 MB  | ##################################### | 100% 
libuuid-2.32.1       | 28 KB     | ##################################### | 100% 
defusedxml-0.7.1     | 23 KB     | ##################################### | 100% 
libstdcxx-ng-9.3.0   | 4.0 MB    | ##################################### | 100% 
ipython_genutils-0.2 | 21 KB     | ##################################### | 100% 
async_generator-1.10 | 18 KB     | ##################################### | 100% 
xorg-libxdmcp-1.1.3  | 19 KB

_pytorch_select-0.2  | 2 KB      | ##################################### | 100% 
jupyterlab_pygments- | 8 KB      | ##################################### | 100% 
libgcc-ng-9.3.0      | 7.8 MB    | ##################################### | 100% 
pillow-8.1.2         | 668 KB    | ##################################### | 100% 
widgetsnbextension-3 | 1.8 MB    | ##################################### | 100% 
backports.functools_ | 8 KB      | ##################################### | 100% 
bokeh-2.3.0          | 8.3 MB    | ##################################### | 100% 
pthread-stubs-0.4    | 5 KB      | ##################################### | 100% 
parso-0.8.1          | 67 KB     | ##################################### | 100% 
py-xgboost-1.3.3     | 124 KB    | ##################################### | 100% 
pysocks-1.7.1        | 27 KB     | ##################################### | 100% 
openjpeg-2.4.0       | 525 KB    | ##################################### | 100% 
scikit-learn-0.24.1  | 7.5 M

cffi-1.14.5          | 225 KB    | ##################################### | 100% 
secretstorage-3.3.1  | 24 KB     | ##################################### | 100% 
ibm-wsrt-py37main-ke | 2 KB      | ##################################### | 100% 
gmp-6.2.1            | 806 KB    | ##################################### | 100% 
boto3-1.17.23        | 70 KB     | ##################################### | 100% 
libprotobuf-3.15.5   | 2.5 MB    | ##################################### | 100% 
patsy-0.5.1          | 187 KB    | ##################################### | 100% 
cached_property-1.5. | 11 KB     | ##################################### | 100% 
gstreamer-1.18.3     | 2.0 MB    | ##################################### | 100% 
oauthlib-3.0.1       | 82 KB     | ##################################### | 100% 
ipykernel-5.5.0      | 166 KB    | ##################################### | 100% 
ibm-wsrt-py37main-ma | 2 KB      | ##################################### | 100% 
cryptography-3.4.4   | 636 K

done


In [26]:

# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

In [27]:
CLIENT_ID = 'HWHXN0CP0ISF0R5C3RPDMKZJIXNFXDISMFFQJUJ0413PLKOP' 
CLIENT_SECRET = 'LL0KG5GZDJ1152MAFPP12YFWSPLMMAWDWTLUYYDEDIIQOAQY'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: HWHXN0CP0ISF0R5C3RPDMKZJIXNFXDISMFFQJUJ0413PLKOP
CLIENT_SECRET:LL0KG5GZDJ1152MAFPP12YFWSPLMMAWDWTLUYYDEDIIQOAQY


In [28]:
#create a function to get all the venue categories in Toronto

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

In [29]:
#Collecting the venues in Toronto for each Neighbourhood

venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [30]:
venues_in_toronto.shape

(1328, 5)

In [31]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3               Portugril  Portuguese Restaurant  
4             Tim Hortons            Coffee Shop

In [32]:
#Venues based on Neighbourhood

venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1314  Mimico NW, The Queensway West, South of Bloor,...   
1315  Mimico NW, The Queensway West, South of Bloor,...   
1316  Mimico NW, The Queensway West, South of Bloor,...   
1317  Mimico NW, The Queensway West, South of Bloor,...   
1318  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572               Portugril   
4                  43.725882               -79.315572             Tim Hortons   
...                      ...                      ...                     ...   
1314               43.628841               -79.520999              Wingporium   
1315               43.628841               -79.520999        South St. Burger   
1316               43.628841               -79.520999               Dollarama   
1317               43.628841               -79.520999          Healthy Planet   
1318               43.628841               -79.520999                7-Eleven   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3     Portuguese Restaurant  
4               Coffee Shop  
...                     ...  
1314            Wings Joint  
1315           Burger Joint  
1316         Discount Store  
1317        Supplement Shop  
1318      Convenience Store  

[416 rows x 5 columns]

In [33]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                     Toronto Dominion Centre, Design Exchange   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.379817   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                          The National Club  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[231 rows x 4 columns]

One Hot encoding the venue Categories

In [34]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
1323                  0        0                   0             0   
1324                  0        0                   0             0   
1325                  0        0                   0             0   
1326                  0        0                   0             0   
1327                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1323               0                0                 0                    0   
1324               0                0                 0                    0   
1325               0                0                 0                    0   
1326               0                0                 0                    0   
1327               0                0                 0                    0   

      Antique Shop  Art Gallery  ...  Train Station  \
0                0            0  ...              0   
1                0            0  ...              0   
2                0            0  ...              0   
3                0            0  ...              0   
4                0            0  ...              0   
...            ...          ...  ...            ...   
1323             0            0  ...              0   
1324             0            0  ...              0   
1325             0            0  ...              0   
1326             0            0  ...              0   
1327             0            0  ...              0   

      Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                                 0                 0            0   
1                                 0                 0            0   
2                                 0                 0            0   
3                                 0                 0            0   
4                                 0                 0            0   
...                             ...               ...          ...   
1323                              0                 0            0   
1324                              0                 0            0   
1325                              0                 0            0   
1326                              0                 0            0   
1327                              0                 0            0   

      Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                         0                0         0            0   
1                         0                0         0            0   
2                         0                0         0            0   
3                         0                0         0            0   
4                         0                0         0            0   
...                     ...              ...       ...          ...   
1323                      0                0         0            0   
1324                      0                0         0            0   
1325                      0                0         0         

In [35]:
#adding neighbourhood to encoded df

toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  \
0                    0             0  ...              0   
1                    0             0  ...              0   
2                    0             0  ...              0   
3                    0             0  ...              0   
4                    0             0  ...              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 232 columns]

In [36]:
#calculate the mean venue categories in each Neighbourhood

toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Train Station  \
0               0.0                 0.00           0.0  ...            0.0   
1               0.0                 0.00           0.0  ...            0.0   
2               0.0                 0.00           0.0  ...            0.0   
3               0.0                 0.00           0.0  ...            0.0   
4               0.0                 0.04           0.0  ...            0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 232 columns]

In [37]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [38]:
import numpy as np

In [39]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant        Clothing Store   
1          Skating Rink                Coffee Shop                   Pub   
2                  Bank                Gas Station            Restaurant   
3    Chinese Restaurant                       Bank   Japanese Restaurant   
4    Italian Restaurant                Coffee Shop      Greek Restaurant   

  5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
0          Skating Rink     Accessories Store                     Motel   
1        Sandwich Place                   Gym     Vietnamese Restaurant   
2           Supermarket   Fried Chicken Joint        Frozen Yogurt Shop   
3     Accessories Store        Medical Center  Mediterranean Restaurant   
4              Pharmacy                  Café                Restaurant   

        8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Mediterranean Restaurant           Men's Store          Metro Station  
1  Modern European Restaurant                Market    Martial Arts School  
2           Mobile Phone Shop                 Diner           Intersection  
3                 Men's Store         Metro Station     Mexican Restaurant  
4                         Pub           Pizza Place         Cosmetics Shop

In [42]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [43]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [44]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1,
       2, 1, 1, 1, 1, 1, 3, 1, 3, 1, 1, 1, 1, 1, 4, 1, 1, 2, 1, 1, 1, 2,
       1, 1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       2, 1, 2, 1, 1, 1, 1, 2, 1], dtype=int32)

In [45]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [46]:
#Join toronto_grouped with combined_data on neighbourhood to add latitude & longitude for each neighborhood to prepare it for plotting

toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels   1st Most Common Venue  \
0  43.753259 -79.329656             2.0                    Park   
1  43.725882 -79.315572             1.0             Pizza Place   
2  43.654260 -79.360636             1.0             Coffee Shop   
3  43.718518 -79.464763             1.0  Furniture / Home Store   
4  43.662301 -79.389494             1.0             Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0     Food & Drink Shop     Accessories Store                 Motel   
1           Coffee Shop          Hockey Arena     French Restaurant   
2                  Park        Breakfast Spot               Theater   
3        Clothing Store     Accessories Store           Event Space   
4                   Gym      Sushi Restaurant                 Diner   

   5th Most Common Venue  6th Most Common Venue       7th Most Common Venue  \
0    Martial Arts School         Massage Studio              Medical Center   
1  Portuguese Restaurant           Intersection  Financial or Legal Service   
2                   Café                 Bakery              Farmers Market   
3               Boutique  Vietnamese Restaurant                 Coffee Shop   
4            Yoga Studio                Theater               Burrito Place   

      8th Most Common Venue      9th Most Common Venue  \
0  Mediterranean Restaurant                Men's Store   
1               Men's Store   Mediterranean Restaurant   
2            Chocolate Shop               Dessert Shop   
3        Miscellaneous Shop  Middle Eastern Restaurant   
4        Mexican Restaurant                       Café   

       10th Most Common Venue  
0               Metro Station  
1              Medical Center  
2         Distribution Center  
3  Modern European Restaurant  
4                    Creperie

In [47]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

In [48]:
#plotting

import matplotlib.cm as cm
import matplotlib.colors as colors

In [49]:

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Cluster 1

In [52]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             0.0                Bakery     Accessories Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11                 Motel   Martial Arts School        Massage Studio   

   6th Most Common Venue     7th Most Common Venue 8th Most Common Venue  \
11        Medical Center  Mediterranean Restaurant           Men's Store   

   9th Most Common Venue 10th Most Common Venue  
11         Metro Station     Mexican Restaurant

Cluster 2

In [51]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels   1st Most Common Venue  \
1          North York             1.0             Pizza Place   
2    Downtown Toronto             1.0             Coffee Shop   
3          North York             1.0  Furniture / Home Store   
4    Downtown Toronto             1.0             Coffee Shop   
6         Scarborough             1.0    Fast Food Restaurant   
..                ...             ...                     ...   
97   Downtown Toronto             1.0             Coffee Shop   
98          Etobicoke             1.0                   River   
99   Downtown Toronto             1.0             Coffee Shop   
100      East Toronto             1.0             Pizza Place   
102         Etobicoke             1.0          Hardware Store   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1             Coffee Shop          Hockey Arena     French Restaurant   
2                    Park        Breakfast Spot               Theater   
3          Clothing Store     Accessories Store           Event Space   
4                     Gym      Sushi Restaurant                 Diner   
6                   Motel                Market   Martial Arts School   
..                    ...                   ...                   ...   
97                   Café            Restaurant    Seafood Restaurant   
98      Accessories Store         Movie Theater   Martial Arts School   
99     Salon / Barbershop      Sushi Restaurant            Restaurant   
100         Garden Center  Fast Food Restaurant            Smoke Shop   
102        Discount Store         Burrito Place     Convenience Store   

     5th Most Common Venue   6th Most Common Venue  \
1    Portuguese Restaurant            Intersection   
2                     Café                  Bakery   
3                 Boutique   Vietnamese Restaurant   
4              Yoga Studio                 Theater   
6           Massage Studio          Medical Center   
..                     ...                     ...   
97                   Hotel  Gluten-free Restaurant   
98          Massage Studio          Medical Center   
99                Beer Bar              Steakhouse   
100          Burrito Place              Skate Park   
102            Flower Shop         Supplement Shop   

          7th Most Common Venue 8th Most Common Venue  \
1    Financial or Legal Service           Men's Store   
2                Farmers Market        Chocolate Shop   
3                   Coffee Shop    Miscellaneous Shop   
4                 Burrito Place    Mexican Restaurant   
6      Mediterranean Restaurant           Men's Store   
..                          ...                   ...   
97                   Steakhouse                   Pub   
98     Mediterranean Restaurant           Men's Store   
99          Martial Arts School             Bookstore   
100          Light Rail Station            Restaurant   
102              Sandwich Place         Tanning Salon   

         9th Most Common Venue      10th Most Common Venue  
1     Mediterranean Restaurant              Medical Center  
2                 Dessert Shop         Distribution Center  
3    Middle Eastern Restaurant  Modern European Restaurant  
4                         Café                    Creperie  
6                Metro Station          Mexican Restaurant  
..                         ...                         ...  
97                    Tea Room                      Bakery  
98               Metro Station          Mexican Restaurant  
99              Breakfast Spot                   Juice Bar  
100             Farmers Market                      Garden  
102                     Bakery        Fast Food Restaurant  

[87 rows x 12 columns]

Cluster 3

In [53]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             2.0                  Park   
21              York             2.0                  Park   
35         East York             2.0                  Park   
52        North York             2.0                  Park   
64              York             2.0                  Park   
66        North York             2.0                  Park   
77         Etobicoke             2.0                  Park   
91  Downtown Toronto             2.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0      Food & Drink Shop     Accessories Store                 Motel   
21         Women's Store                  Pool     Accessories Store   
35     Convenience Store         Metro Station     Accessories Store   
52     Accessories Store                 Motel   Martial Arts School   
64         Jewelry Store     Accessories Store                 Motel   
66     Convenience Store     Accessories Store                 Motel   
77     Mobile Phone Shop        Sandwich Place     Accessories Store   
91            Playground                 Trail     Accessories Store   

   5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
0    Martial Arts School        Massage Studio            Medical Center   
21                 Motel   Martial Arts School            Massage Studio   
35                 Motel   Martial Arts School            Massage Studio   
52        Massage Studio        Medical Center  Mediterranean Restaurant   
64   Martial Arts School        Massage Studio            Medical Center   
66   Martial Arts School        Massage Studio            Medical Center   
77                 Motel   Martial Arts School            Massage Studio   
91                 Motel   Martial Arts School            Massage Studio   

       8th Most Common Venue     9th Most Common Venue 10th Most Common Venue  
0   Mediterranean Restaurant               Men's Store          Metro Station  
21            Medical Center  Mediterranean Restaurant            Men's Store  
35            Medical Center  Mediterranean Restaurant            Men's Store  
52               Men's Store             Metro Station     Mexican Restaurant  
64  Mediterranean Restaurant               Men's Store          Metro Station  
66  Mediterranean Restaurant               Men's Store          Metro Station  
77            Medical Center  Mediterranean Restaurant            Men's Store  
91            Medical Center  Mediterranean Restaurant            Men's Store

Cluster 4

In [54]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
32      Scarborough             3.0            Playground   
83  Central Toronto             3.0            Playground   
85      Scarborough             3.0            Playground   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
32        Cosmetics Shop     Accessories Store                 Motel   
83     Accessories Store         Movie Theater   Martial Arts School   
85                  Park          Intersection     Accessories Store   

   5th Most Common Venue 6th Most Common Venue     7th Most Common Venue  \
32   Martial Arts School        Massage Studio            Medical Center   
83        Massage Studio        Medical Center  Mediterranean Restaurant   
85                 Motel        Massage Studio            Medical Center   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
32  Mediterranean Restaurant           Men's Store          Metro Station  
83               Men's Store         Metro Station     Mexican Restaurant  
85  Mediterranean Restaurant           Men's Store          Metro Station

In [55]:
#cluster 5
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             4.0        Baseball Field     Accessories Store   
101   Etobicoke             4.0        Baseball Field          Home Service   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57                  Motel   Martial Arts School        Massage Studio   
101     Accessories Store                 Motel   Martial Arts School   

    6th Most Common Venue     7th Most Common Venue     8th Most Common Venue  \
57         Medical Center  Mediterranean Restaurant               Men's Store   
101        Massage Studio            Medical Center  Mediterranean Restaurant   

    9th Most Common Venue 10th Most Common Venue  
57          Metro Station     Mexican Restaurant  
101           Men's Store          Metro Station

Toronto Neighbourhood Clustered based on venue categories